pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,360 @@
+// converge the two. It then updates the Status block of the ManagedCluster resource
+// with the current status of the resource.
+// Duration <=0 means don't need to requeue this obj.
+func (c *Controller) syncHandler(key string) (time.Duration, error) {
+	// If an error occurs during handling, we'll requeue the item so we can
+	// attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+
+	// Convert the namespace/name string into a distinct namespace and name
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return time.Duration(0), nil
+	}
+
+	klog.V(4).Infof("start processing ManagedCluster %q", key)
+	// Get the ManagedCluster resource with this name
+	mcls, err := c.clusterLister.ManagedClusters(ns).Get(name)
+	// The ManagedCluster resource may no longer exist, in which case we stop processing.
+	if apierrors.IsNotFound(err) {
+		klog.V(2).Infof("ManagedCluster %q has been deleted", key)
+		return time.Duration(0), nil
+	}
+	if err != nil {
+		return time.Duration(0), err
+	}
+
+	err = c.updateClusterReadyConditionIfNeed(c.ctx, mcls, retry.DefaultBackoff)
+	if err != nil {
+		return time.Duration(0), err
+	}
+	// requeue duration is a little longer than the duration we consider a cluster to be Unknown
+	requeueAfter := time.Second*time.Duration(*mcls.Status.HeartbeatFrequencySeconds)*ClusterNotReadyThreshold + 10*time.Second
+	return requeueAfter, nil
+}
+
+// enqueue takes a ManagedCluster resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than ManagedCluster.
+func (c *Controller) enqueue(cluster *clusterapi.ManagedCluster) {
+	key, err := cache.MetaNamespaceKeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+func (c *Controller) updateClusterReadyConditionIfNeed(ctx context.Context, cluster *clusterapi.ManagedCluster, backoff wait.Backoff) error {
+	currentReadyCondition := GetClusterCondition(&cluster.Status, clusterapi.ClusterReady)
+	observedReadyCondition := c.getClusterReadyCondition(cluster)
+
+	if observedReadyCondition != nil && !equality.Semantic.DeepEqual(currentReadyCondition, observedReadyCondition) {
+		// update Readyz,Livez,Healthz to false when cluster status is unknown
+		if observedReadyCondition.Status == metav1.ConditionUnknown {
+			cluster.Status.Readyz = false
+			cluster.Status.Livez = false
+			cluster.Status.Healthz = false
+		}
+
+		cluster.Status.Conditions = []metav1.Condition{*observedReadyCondition}
+
+		// retry update due to the network delay or update conflict
+		var lastError error
+		var mcls *clusterapi.ManagedCluster
+		err := wait.ExponentialBackoffWithContext(ctx, backoff, func() (done bool, err error) {
```

Comment:
Will the total time of these retries exceed grace period?

Reply:
Mostly not, but we can't totally avoid it unless return error and requeue to implement retry.

Reply:
Just return the error and requeue.